### PR TITLE
Exclusion filter for project listing

### PR DESCRIPTION
**Pull request by mdeniz**
Pair Programming with @DavidKang

### CODE REVIEW

**Comment by adrianschroeter** [comment]
I don't like the "exclude_projects" term so much, it isn't obvious that this is about the default list of webui in first place. In the backend context it sounds like it would not build them.

What about "unlisted_projects" instead? @evanrolfe can you come up with a better term as native speaker? ;)

File: src/api/app/models/project.rb [hunk 1]
```diff
@@ -100,6 +100,7 @@ def autocomplete(search)
   scope :maintenance_release, -> { where("kind = 'maintenance_release'") }
   scope :home, -> { where("name like 'home:%'") }
   scope :not_home, -> { where.not("name like 'home:%'") }
+  scope :excluded, -> { where.not("name rlike ?", ::Configuration.excluded_projects) if ::Configuration.excluded_projects.present? }
```
Comment (ChrisBr) [inline src/api/app/models/project.rb]: JFYI: ``rlike`` does not use MySQL index. However, probably hard to rewrite to something more performant if we don't know what the OBS Admins will do with it ....
https://www.psce.com/blog/2012/03/31/mysql-queries-with-regexp/

Comment (mdeniz) [inline src/api/app/models/project.rb]: We can switch to LIKE, but then the admin need to write SQL for the filter

Comment (evanrolfe) [inline src/api/app/models/project.rb]: The name `scope :excluded` seems like it means "projects that are excluded" which isn't what this scope is giving. Maybe call it `scope :filtered_for_index`? Or just `filtered`?

File: docs/api/api/configuration.rng [hunk 2]
```diff
@@ -198,6 +198,18 @@
             </choice>
           </element>
         </optional>
+        <optional>
+          <element name="excluded_projects">
```
Comment (evanrolfe) [inline docs/api/api/configuration.rng]: Maybe call it `projects_exclusion_regex`?

File: docs/api/api/configuration.rng [hunk 3]
```diff
@@ -198,6 +198,18 @@
...
+            <!-- The exclusion filter itself. Is a Regular Expression (eg: ^home:.*) -->
+            <text/>
+          </element>
+        </optional>
+        <optional>
+          <element name="excluded_projects_description">
```
Comment (evanrolfe) [inline docs/api/api/configuration.rng]: Maybe `projects_exclusion_regex_description`? Or  `projects_exclusion_description`? Just to keep it consistent with the previous config var..

File: src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+class AddExcludeProjectsToConfiguration < ActiveRecord::Migration[5.0]
+  def self.up
+    add_column :configurations, :excluded_projects, :string
+    add_column :configurations, :excluded_projects_description, :string
+
+    # update database from file
+    old = CONFIG['global_write_through']
+    CONFIG['global_write_through'] = false
+    ::Configuration.first.update_from_options_yml
+    CONFIG['global_write_through'] = old
```
Comment (evanrolfe) [inline src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb]: This line is a bit confusing to me, I don't see whats the point of setting this constant here?

Comment (mdeniz) [inline src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb]: CONFIG['global_write_through'] controls the write (or not) to the backend, so, we disable it before updating the config for only writing in the DB and then we let it as it was before

Comment (hennevogel) [inline src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb]: Why?

**Comment by mdeniz** [comment]
Thanks for the feedback @adrianschroeter and @evanrolfe, finally we are going to use `filtered` for the scope "excluded", `projects_exclusion_filter` for "excluded_projects" and `projects_exclusion_filter_description` for "excluded_projects_description"

**Comment by mdeniz** [comment]
Thanks for the feedback @adrianschroeter and @evanrolfe, finally we are going to use `filtered_for_list` for the scope "excluded", `unlisted_projects_filter` for "excluded_projects" and `unlisted_projects_filter_description` for "excluded_projects_description"

**Comment by mdeniz** [comment]
@evanrolfe if an inclusion filter regexp is needed the admin only needs to negate the regexp... 😄 , but the usual use case is excluding, so we cover both.

File: src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb (hunk 4 above)
Comment (evanrolfe) [inline src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb]: :+1:

**Comment by mdeniz** [comment]
[(2) P2: Configurable project listing filter](https://trello.com/c/DfLO68U1/162-2-p2-configurable-project-listing-filter)

File: src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb (hunk 4 above)
Comment (mdeniz) [inline src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb]: @hennevogel Why what? you mean why we disable the writing or why we have to recover the old state?

Comment (hennevogel) [inline src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb]: Why do you disable the writing?

Comment (mdeniz) [inline src/api/db/migrate/20170111114943_add_exclude_projects_to_configuration.rb]: We saw it here https://github.com/openSUSE/open-build-service/blob/master/src/api/db/migrate/20130702083665_options_yml_to_configuration.rb#L23 and thought that maybe is needed... but taking a look at it right now I don't have it clear. Maybe @adrianschroeter remember why it was needed then 😃

**Comment by adrianschroeter** [comment]
On Donnerstag, 12. Januar 2017, 06:20:02 CET wrote Moisés Déniz Alemán:
> mdeniz commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,17 @@
> +class AddExcludeProjectsToConfiguration < ActiveRecord::Migration[5.0]
> +  def self.up
> +    add_column :configurations, :excluded_projects, :string
> +    add_column :configurations, :excluded_projects_description, :string
> +
> +    # update database from file
> +    old = CONFIG['global_write_through']
> +    CONFIG['global_write_through'] = false
> +    ::Configuration.first.update_from_options_yml
> +    CONFIG['global_write_through'] = old
> 
> We saw it here https://github.com/openSUSE/open-build-service/blob/master/src/api/db/migrate/20130702083665_options_yml_to_configuration.rb#L23 and thought that maybe is needed... but taking a look at it right now I don't have it clear. Maybe @adrianschroeter remember why it was needed then 😃 


migrations need to work without a source backend server running.
Therefore we write to the backend via a delayed job after migration,
once the system is booting up.

You need a 

 Delayed::Job.enqueue WriteConfigurationJob.new


or rename the 20130702203665_write_configuration.rb migration to trigger
that.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by mdeniz** [comment]
So, no need to do it... I will remove then that

**Comment by adrianschroeter** [comment]
I do not see where the default filter to home:.+  is set when migrating or for new installtions. I see it only in test cases and fixtures. Have I missed it?

**Comment by DavidKang** [comment]
@adrianschroeter, the default value is loaded with the option.yml when we run the migration into the database (not into the backend).

**Comment by hennevogel** [comment]
@DavidKang but if I already have an `option.yml` it won't be added. So this works for new installations, not for updates.

**Comment by mdeniz** [comment]
@hennevogel Done, we added a default value for the filters in the migration

**Comment by adrianschroeter** [comment]
Is there any reason why we have two places to set it? /configuration and options.yml?

The code which is reading it from options.yml and stores it to DB is for backward compatibility, but is there a reason to have that for new settings?

**Comment by mdeniz** [comment]
Ok @adrianschroeter , we removed those new params from the options.yml, so they only live in the DB.

**Comment by mdeniz** [comment]
@hennevogel do you mind to re-review it? 😄

File: src/api/spec/controllers/webui/configuration_controller_spec.rb [hunk 5]
```diff
@@ -17,4 +17,22 @@
       end
     end
   end
+
+  describe 'PATCH #update' do
+    context 'as admin' do
+      before do
+        login(admin_user)
+        patch :update, configuration: { name: 'obs', title: 'OBS', description: 'something',
```
Comment (hennevogel) [inline src/api/spec/controllers/webui/configuration_controller_spec.rb]: Please test the default values...

Comment (mdeniz) [inline src/api/spec/controllers/webui/configuration_controller_spec.rb]: Done, I use other values, not the default ones, so we can see that the update action works

**Comment by mdeniz** [comment]
@hennevogel do you mind to review it again? 😸

**Pull request merged**